### ISSUE TITLE
[ibm-telco] Horizontal Scale up of worker node using ztp playbooks 

### ISSUE THREAD

**Issue by pmallan** [redhat-ztp/ztp-cluster-deploy#100]
Hi @yrobla we have question on scaling up worker using ztp playbook. Is it possible and supported screnario?

    Control Plane Cluster and a worker node has been deployed using ztp play book
    Try to add another worker node to the same control plane cluster using playbook. We tried this and notice that already existing worker node gets reinstalled. The expectation is that, add new worker node to the cluster without impacting existing worker node.
    We can deploy multiple worker nodes to the cluster during first time deployment however can not scale up later.
    Please update us whether it has been tested and supported scenario or not.
We have deployed cluster with one worker node.

[root@vcuhost ~]# aicli list clusters
+----------+--------------------------------------+--------------+---------------------+
| Cluster  |                  Id                  |    Status    |      Dns Domain     |
+----------+--------------------------------------+--------------+---------------------+
|   vcu    | d2571b9b-6e33-4122-a274-98dd97db0563 |  installed   | rtp.raleigh.ibm.com |
| vcu-day2 | b533cfe0-63c7-4936-9039-b4d2da04a3bf | adding-hosts |         N/A         |
+----------+--------------------------------------+--------------+---------------------+
[root@vcuhost ~]# aicli list hosts
+--------------+----------+--------------------------------------+---------------------------+--------+----------------+
|     Host     | Cluster  |                  Id                  |           Status          |  Role  |       Ip       |
+--------------+----------+--------------------------------------+---------------------------+--------+----------------+
| vcu-master-0 |   vcu    | d7ece5e2-64b5-4f0a-ad62-7b0155923fdd |         installed         | master | 172.29.231.142 |
| vcu-master-1 |   vcu    | 9f8b8d9a-960b-4b17-8997-11341fede901 |         installed         | master | 172.29.231.143 |
| vcu-master-2 |   vcu    | 632b3b85-e30e-4ddb-8e6e-73784d1ea716 |         installed         | master | 172.29.231.144 |
| vcu-worker-1 | vcu-day2 | 1c907474-713e-6617-ff02-a68438e3de04 | added-to-existing-cluster | worker |      N/A       |
+--------------+----------+--------------------------------------+---------------------------+--------+----------------+

You will see that in inventory/hosts.sample there is a "redeploy=False" flag. If you set it to the nodes that have already been deployed, the system will skip the provision of those.

**Comment by pmallan** [comment]
@yrobla 

"redeploy=false" is not referenced anywhere in the playbooks... I even verified the playbooks going inside.
It looks like a simple check to be introduced, but currently missing.

[root@vcuhost ztp-cluster-deploy]# grep -r redeploy
ai-deploy-cluster-remoteworker/inventory/hosts.sample:worker-0.test-aut.cluster-testing name=worker_0 bmc_type=SuperMicro bmc_address=192.168.111.212 bmc_user="ADMIN" bmc_password="ADMIN" smb_host=192.168.111.1 smb_path=share ip=<host_ip> hostname=<hostname> redeploy=false profile=worker-cnf
[root@vcuhost ztp-cluster-deploy]# git branch
* master

Please suggest whether AI support this scaling of worker node or not

**Comment by yrobla** [comment]
Sure, the problem was an accidental drop of the feature due to some refactor. I have re-added the feature again, so you should be able to use it now.

**Comment by maheshd2** [comment]
@yrobla As we found redeploy flag is not implemented in the playbooks and also we found its quite tricky to use redeploy=true/false for the worker node addition via automation. 

So, we modified the code as follows to make it work. We added this logic to unblock ourselves and test the day2 worker node addition feature. Please have a look and provide us valid solution.

`./ai-deploy-cluster-remoteworker/roles/add-remote-workers/tasks/main.yml`
```
---
- name: Enroll a worker node
  include_tasks: enroll.yaml
  with_items: "{{ groups['worker_nodes'] }}"

- name: Wait until hosts are ready in Assisted Installer
  shell: "aicli list hosts | grep {{ cluster_name }}-day2 | grep 'insufficient\\|known\\|added-to-existing-cluster' | wc -l"
  register: hosts_ready
  until: hosts_ready.stdout|int == groups['worker_nodes'] | length
  retries: 100
  delay: 30
  environment:
    AI_URL: "{{ ai_url }}"
```

`./ai-deploy-cluster-remoteworker/roles/add-remote-workers/tasks/enroll.yaml`
```
---
- name: Check if worker node already added to cluster
  shell: "aicli list hosts | grep {{ hostvars[item].name }} | grep 'added-to-existing-cluster' | wc -l"
  register: host_added
  environment:
    AI_URL: "{{ ai_url }}"

- name: Enroll a SuperMicro node
  include_role:
    name: "{{ playbook_dir }}/../common-roles/enroll-supermicro"
  vars:
    iso_name: "{{ cluster_name }}-day2.iso"
    smb_host: "{{ hostvars[item].smb_host }}"
    smb_path: "{{ hostvars[item].smb_path }}"
    bmc_address: "{{ hostvars[item].bmc_address }}"
    bmc_user: "{{ hostvars[item].bmc_user }}"
    bmc_password: "{{ hostvars[item].bmc_password }}"
  when: hostvars[item].bmc_type == "SuperMicro" and host_added.stdout|int == 0

- name: Enroll a Dell node
  include_role:
    name: "{{ playbook_dir }}/../common-roles/enroll-dell"
  vars:
    iso_name: "{{ cluster_name }}-day2.iso"
    bmc_user: "{{ hostvars[item].bmc_user }}"
    bmc_password: "{{ hostvars[item].bmc_password }}"
    bmc_address: "{{ hostvars[item].bmc_address }}"
  when: hostvars[item].bmc_type == "Dell" and host_added.stdout|int == 0
```

**Comment by maheshd2** [comment]
> 
> 
> Sure, the problem was an accidental drop of the feature due to some refactor. I have re-added the feature again, so you should be able to use it now.

This change has introduced new issue, https://github.com/redhat-ztp/ztp-cluster-deploy/commit/9c42ec585c8b0ab6f7b9c2d1438779b99c0af33c#commitcomment-49734073

**Comment by maheshd2** [comment]
@yrobla Yesterday we noticed one more issue with day2 worker node addition w.r.t below task.

File Name: `./ai-deploy-cluster-remoteworker/roles/add-remote-workers/tasks/main.yml`
```
- name: Wait until hosts are ready in Assisted Installer
  shell: "aicli list hosts | grep {{ cluster_name }}-day2 | grep 'insufficient\\|known' | wc -l"
  register: hosts_ready
  until: hosts_ready.stdout|int == groups['worker_nodes'] | length
  retries: 100
  delay: 30
  environment:
    AI_URL: "{{ ai_url }}"

```
Since some of the nodes will be already deployed, the state of the deployed node will be 'added-to-existing-cluster', Hence we have to include hosts with this state as well in the below task.
```
- name: Wait until hosts are ready in Assisted Installer
  shell: "aicli list hosts | grep {{ cluster_name }}-day2 | grep 'insufficient\\|known\\|added-to-existing-cluster' | wc -l"
  register: hosts_ready
  until: hosts_ready.stdout|int == groups['worker_nodes'] | length
  retries: 100
  delay: 30
  environment:
    AI_URL: "{{ ai_url }}"
```

Otherwise the day2 worker node addition will be struck/fail at this stage.
![image](https://user-images.githubusercontent.com/27758674/115421413-8be73680-a219-11eb-85e8-4d3fb22347e7.png)

**Comment by pmallan** [comment]
@yrobla we tried with latest code from stable/4.7. The issue still persist. Unable to add additional worker node 
[root@ztpautoqa ~]# aicli list hosts
+---------------------+----------------+--------------------------------------+---------------------------+-------------+-------------+
|         Host        |    Cluster     |                  Id                  |           Status          |     Role    |      Ip     |
+---------------------+----------------+--------------------------------------+---------------------------+-------------+-------------+
| ztpauto-master-qa-1 |   ztpautoqa    | ca71b022-6f55-4664-9ece-563cfceaa0ba |         installed         |    master   | 9.42.98.153 |
| ztpauto-master-qa-2 |   ztpautoqa    | c938c434-3513-4c94-83e8-37bd68e425c7 |         installed         |    master   | 9.42.98.154 |
| ztpauto-master-qa-3 |   ztpautoqa    | 6313a808-2cb5-48e2-b9c5-de51d173a93f |         installed         |    master   | 9.42.98.155 |
| ztpauto-worker-qa-2 | ztpautoqa-day2 | 92e02726-2a00-3d52-ca05-7d0082c4b723 | added-to-existing-cluster |    worker   |     N/A     |
| ztpauto-worker-qa-3 | ztpautoqa-day2 | c4e31ff7-945b-98b8-f32d-66af58adbfc8 |           known           | auto-assign |     N/A     |
+---------------------+----------------+--------------------------------------+---------------------------+-------------+-------------+

<img width="1566" alt="Screenshot 2021-04-29 at 4 10 53 PM" src="https://user-images.githubusercontent.com/79190301/116544230-7a361b00-a90c-11eb-97b5-da7856f2a425.png">

**Comment by pmallan** [comment]
@yrobla any update on this? We would like to complete the ZTP work asap.

**Comment by yrobla** [comment]
This should be fixed now, i merged a commit today that will fix the problem. Can you try with latest version?
